File: services/recommendation.service.js
import { spawn } from "child_process";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function getMLRecommendations(preferences, allPlaces, destination) {
    return new Promise((resolve, reject) => {
        // console.log("🧠 Calling ML Engine for Recommendations...");

        const pyScript = path.join(__dirname, "../ml_engine/run_recommendations.py");
        const pythonProcess = spawn("python3", [pyScript]);

        let dataString = "";
        let errorString = "";

        // Prepare payload
        // We send preferences and destination.
        const payload = JSON.stringify({ preferences, destination });

        pythonProcess.stdin.write(payload);
        pythonProcess.stdin.end();

        pythonProcess.stdout.on("data", (data) => {
            dataString += data.toString();
        });

        pythonProcess.stderr.on("data", (data) => {
            errorString += data.toString();
        });

        pythonProcess.on("close", (code) => {
            if (code !== 0) {
                console.error(`ML script exited with code ${code}`);
                console.error(errorString);
                resolve([]); // Fallback to empty (will trigger heuristic fallback)
            } else {
                try {
                    const results = JSON.parse(dataString);
                    if (results.error) {
                        console.error("ML Engine Error:", results.error);
                        resolve([]);
                    } else {
                        // console.log(`✅ ML returned ${results.length} recommendations`);
                        resolve(results);
                    }
                } catch (e) {
                    console.error("Failed to parse ML response:", e);
                    resolve([]);
                }
            }
        });
    });
}

File: services/recommendation_service.py
"""
Recommendation Service - Wrapper for ML-based recommendations
"""
import os
import json
import subprocess
import logging

logger = logging.getLogger(__name__)

# Path to ML script
ML_SCRIPT_PATH = os.path.join(
    os.path.dirname(__file__), 
    "..", 
    "ml_engine", 
    "run_recommendations.py"
)


def get_ml_recommendations(preferences: list, destination: str) -> list:
    """
    Get ML-based recommendations for a destination.
    
    Args:
        preferences: List of user preference strings
        destination: Target destination name
    
    Returns:
        List of recommended places or empty list on error
    """
    try:
        logger.info("🧠 Calling ML Engine for Recommendations...")
        
        # Prepare payload
        payload = json.dumps({
            "preferences": preferences,
            "destination": destination
        })

        # Run Python ML script
        result = subprocess.run(
            ["python3", ML_SCRIPT_PATH],
            input=payload,
            capture_output=True,
            text=True,
            timeout=30
        )

        if result.returncode != 0:
            logger.error(f"ML script exited with code {result.returncode}")
            logger.error(result.stderr)
            return []

        # Parse results
        try:
            results = json.loads(result.stdout)
            if isinstance(results, dict) and results.get("error"):
                logger.error(f"ML Engine Error: {results['error']}")
                return []
            
            logger.info(f"✅ ML returned {len(results)} recommendations")
            return results
            
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse ML response: {e}")
            return []

    except subprocess.TimeoutExpired:
        logger.error("ML script timed out")
        return []
    except Exception as e:
        logger.error(f"Error calling ML engine: {e}")
        return []
